hail/python/hail/matrixtable.py

```diff
@@ -539,6 +539,136 @@ class MatrixTable(ExprContainer):
+        |   int32 | int32 |
+        +---------+-------+
+        |       0 |     2 |
+        |       1 |     3 |
+        +---------+-------+
+        >>> mt.entry.show()
+        +---------+-------------+-------------+
+        | row_idx | <col 0>.baz | <col 1>.baz |
+        +---------+-------------+-------------+
+        |   int32 |       int32 |       int32 |
+        +---------+-------------+-------------+
+        |       0 |           1 |           2 |
+        |       1 |           3 |           4 |
+        +---------+-------------+-------------+
+
+        Parameters
+        ----------
+        globals : :class:`dict` from :class:`str` to :obj:`any`
+            Global fields by name
+
+        rows: :class:`dict` from :class:`str` to :class:`list` of :obj:`any`
+            Row fields by name
+
+        cols: :class:`dict` from :class:`str` to :class:`list` of :obj:`any`
+            Column fields by name
+
+        entries: :class:`dict` from :class:`str` to :class:`list` of :class:`list` of :obj:`any`
+            Matrix entries by name in the form `entry[row_idx][col_idx]`
+
+        Returns
+        -------
+        :class:`.MatrixTable`
+            A MatrixTable assembled from inputs keyed by `row_idx`
+
+        Notes
+        -----
+        - The number of rows and columns specified in `rows` and `cols` for all
+          fields must match the matrix dimensions in `entries`.
+        - This method is intended for educational and testing purposes only.
+        """
+        # General idea: build a `Table` representation matching that returned by
+        # `MatrixTable.localize_entries` and then call `_unlocalize_entries`.
+        def invert(kvs: Dict[str, Iterable[Any]]) -> List[Dict[str, Any]]:
+            return [dict(zip(kvs, vs)) for vs in zip(*kvs.values())]
+
+        def anyval(kvs):
+            return next(iter(kvs.values()))
+
+        # In the case rows or cols aren't specified, we need to infer the
+        # matrix dimensions from *an* entry. Which one isn't important as we
+        # won't enforce congruence among input dimensions.
+        assert not ((rows is None or cols is None) and (entries is None))
+        cols = invert(cols) if cols else [{} for _ in anyval(entries)[0]]
+        for i, _ in enumerate(cols):
+            cols[i] = hl.struct(col_idx=i, **cols[i])
+
+        # While globals aren't required, `MatrixTable._unlocalize_entries`
+        # extracts matrix columns from the the `Table`'s globals.
+        if globals is None:
+            globals = {}
+
+        cols_field_name = 'columns'
+        globals[cols_field_name] = cols
+
+        # `MatrixTable._unlocalize_entries` extracts matrix entries from the
+        # rows on the table, so these need to be defined too.
+        rows = invert(rows) if rows else [{} for _ in anyval(entries)]
+        entries = map(invert, invert(entries)) \
+            if entries \
+            else [[{} for _ in cols] for _ in rows]
```

Comment:
In the spirit of consistency across the codebase, this expression is usually written as:
```
if entries:
    entries = map(invert, invert(entries))
else:
    entries = [[{} for _ in cols] for _ in rows]
```
`mypy` should infer the type just fine.

We generally avoid line continuation in Python.

Reply:
I have a bunch of ternaries in this function so putting this would be weird without changing those too. If the issue is avoiding the line continuation, I'll capitulate on 80 columns.